### PR TITLE
support for blackbox presets

### PR DESCRIPTION
**Pull request by sakitume**
* Update profile_t with new blackbox_preset_t member for persistence and exposure to Configurator
* Blackbox "file" now has a "blackbox_fieldflags" member to indicate what fields were emitted
* OSD menu updated to allow choosing a blackbox preset

### CODE REVIEW

File: src/io/blackbox.h [hunk 1]
```diff
@@ -28,7 +28,29 @@ typedef struct {
...
+  BBOX_FIELD_SETPOINT,
+  BBOX_FIELD_ACCEL_RAW,
+  BBOX_FIELD_ACCEL_FILTER,
+  BBOX_FIELD_GYRO_RAW,
+  BBOX_FIELD_GYRO_FILTER,
+  BBOX_FIELD_MOTOR,
+  BBOX_FIELD_CPU_LOAD,
+  BBOX_FIELD_DEBUG,
+
+  _BBOX_FIELD_MAX,
```
Comment (bkleiner) [inline src/io/blackbox.h:48]: other max entries do not carry a leading `_`, lets remove it here too.

**Review by bkleiner — CHANGES_REQUESTED** [review]
As already mentioned on discord, i really like the general idea here. let's address the code style stuff and this should be good to go.

**Comment by bkleiner** [comment]
ah one more thing, please bump the minor for the `QUIC_PROTOCOL_VERSION` in `quic.h` line 11, we are breaking protocol compatibility. We gonna need that to gate the functionality in the config.

File: src/core/profile.c [hunk 2]
```diff
@@ -13,6 +13,35 @@
...
+        .blackbox_fieldflags =  (1 << BBOX_FIELD_GYRO_FILTER) | (1 << BBOX_FIELD_LOOP) | (1 << BBOX_FIELD_TIME),
+        .rate_divisor = 20,
+        .name = "Gyro Filtered, Rate Divisor=20",
+        .name_osd = "GYRO FILTERED, RATE=20"
+    },
+};
+const uint32_t blackbox_presets_count = sizeof(blackbox_presets) / sizeof(blackbox_preset_t);
+
+void blackbox_preset_apply(const blackbox_preset_t* preset, profile_blackbox_t* profile)
+{
```
[the remark this answers is not in the document]
Comment (sakitume) [reply src/core/profile.c]: As requested, files resaved after applying "Format document"

File: src/io/blackbox.c [hunk 3]
```diff
@@ -9,39 +9,66 @@
...
   CBOR_CHECK_ERROR(cbor_result_t res = cbor_encode_array_indefinite(enc));
 
+  // loop and time are always emitted regardless of what blackbox_fieldflags indicates
   CBOR_CHECK_ERROR(res = cbor_encode_uint32(enc, &b->loop));
   CBOR_CHECK_ERROR(res = cbor_encode_uint32(enc, &b->time));
 
-  CBOR_CHECK_ERROR(res = cbor_encode_compact_vec3_t(enc, &b->pid_p_term));
-  CBOR_CHECK_ERROR(res = cbor_encode_compact_vec3_t(enc, &b->pid_i_term));
-  CBOR_CHECK_ERROR(res = cbor_encode_compact_vec3_t(enc, &b->pid_d_term));
+  if ((1 << BBOX_FIELD_PID_P_TERM) & blackbox_fieldflags) {
```
[the remark this answers is not in the document]
Comment (sakitume) [reply src/io/blackbox.c]: Positions of operands have been swapped as requested

File: src/io/blackbox.h (hunk 1 above)
Comment (sakitume) [reply src/io/blackbox.h]: I wanted to make it clear that BBOX_FIELD_MAX was not simply another field, but I can see how you would prefer consistency. The leading underscore has been removed as requested

**Comment by sakitume** [comment]
I went ahead and bumped up QUIC_PROTOCOL_VERSION as requested, and I agree, the protocol is not identical however I don't think it is actually incompatible in any way. 

I'm curious what I might be missing

**Comment by bkleiner** [comment]
> I went ahead and bumped up QUIC_PROTOCOL_VERSION as requested, and I agree, the protocol is not identical however I don't think it is actually incompatible in any way.
> 
> I'm curious what I might be missing

the blackbox files do not necessarily include all fields anymore. i know that you have a quite clever workaround for the BF converter, but _strictly speaking_ we have removed data from the api surface.

**Review by bkleiner — APPROVED** [review]

**Pull request merged**